validator_worker/src/main.rs

```diff
@@ -98,26 +120,100 @@ fn main() {
     }
 }
 
-// @TODO work in separate pull request
-fn run(_is_single_tick: bool, sentry: &str, config: &Config, adapter: impl Adapter + 'static) {
-    let _sentry_url = sentry.to_owned();
-    let _adapter = adapter;
-    let _config = config.clone();
-
-    // let result = async move {
-    //     let channels = await!(all_channels(&sentry_url, adapter.clone())).unwrap();
-    //     println!("{:?}", channels);
-    //     for channel in channels.into_iter() {
-    //         let sentry = SentryApi::new(adapter.clone(), &channel, &config, true);
-    //         let whoami = adapter.whoami();
-    //         let index = channel.spec.validators.into_iter().position(|v| v.id == whoami);
-    //         let tick = match index {
-    //             Some(0) => Leader.tick(&channel),
-    //             Some(1) => Follower.tick(&channel)
-    //         };
-    //     }
-    //     Ok(())
-    // };
-    // @TODO hanlde errors more gracefully
-    // tokio::run(result.map_err(|e: Box<dyn Error>| panic!("{}", e)).boxed().compat())
+fn run<A: Adapter + 'static>(is_single_tick: bool, sentry_url: &str, config: &Config, adapter: A) {
+    let sentry_adapter = Arc::new(Mutex::new(adapter.clone()));
+    let whoami = adapter.whoami();
+
+    let args = Args {
+        sentry_url: sentry_url.to_owned(),
+        config: config.to_owned(),
+        adapter: sentry_adapter,
+        whoami,
+    };
+
+    if is_single_tick {
+        tokio::run(iterate_channels(args).boxed().compat());
+    } else {
+        tokio::run(infinite(args).boxed().compat());
+    }
+}
+
+async fn infinite<A: Adapter + 'static>(args: Args<A>) -> Result<(), ()> {
+    loop {
+        let arg = args.clone();
+        let delay_future =
+            Delay::new(Instant::now().add(Duration::from_secs(arg.config.wait_time as u64)));
+        let joined = join(iterate_channels(arg), delay_future.compat());
+        match joined.await {
+            (_, Err(e)) => eprintln!("{}", e),
+            _ => println!("finished processing channels"),
+        };
+    }
+}
+
+async fn iterate_channels<A: Adapter + 'static>(args: Args<A>) -> Result<(), ()> {
+    let channels = all_channels(&args.sentry_url, args.whoami.clone())
+        .await
+        .expect("Failed to get channels");
+    let channels_size = channels.len();
+
+    try_join_all(
+        channels.into_iter().map(|channel| {
+            validator_tick(args.adapter.clone(), channel, &args.config, &args.whoami)
+        }),
+    )
+    .await
+    .expect("Failed to iterate channels");
+
+    println!("processed {} channels", channels_size);
+    if channels_size >= args.config.max_channels as usize {
+        println!(
+            "WARNING: channel limit cfg.MAX_CHANNELS={} reached",
+            args.config.max_channels
+        )
+    }
+
+    Ok(())
+}
+
+async fn validator_tick<A: Adapter + 'static>(
+    adapter: Arc<Mutex<A>>,
+    channel: Channel,
+    config: &Config,
+    whoami: &str,
+) -> Result<(), ()> {
```

Comment:
I think you can just remove the return type and not use a Result.

Reply:
That's added because of `try_join_all`